src/client/java/net/neoforged/neoforge/client/neo3d/validation/GpuDeviceUsageValidator.java

```diff
@@ -0,0 +1,134 @@
+            }
+        } else if (depthOrLayers == 1) {
+            // 2d texture
+            if (width > properties.maximumImageDimension2D()) {
+                throw new IllegalArgumentException(String.format("Cannot create 2d texture with width (%d) in excess of 'maximumImageDimension2D' (%d)", width, properties.maximumImageDimension2D()));
+            }
+            if (height > properties.maximumImageDimension2D()) {
+                throw new IllegalArgumentException(String.format("Cannot create 2d texture with height (%d) in excess of 'maximumImageDimension2D' (%d)", height, properties.maximumImageDimension2D()));
+            }
+        } else if ((usage & GpuTexture.USAGE_CUBEMAP_COMPATIBLE) != 0) {
+            // cubemap
+            if (width > properties.maximumImageDimensionCube()) {
+                throw new IllegalArgumentException(String.format("Cannot create cubemap texture with height (%d) in excess of 'maximumImageDimensionCube' (%d)", width, properties.maximumImageDimensionCube()));
+            }
+            if (height > properties.maximumImageDimensionCube()) {
+                throw new IllegalArgumentException(String.format("Cannot create cubemap texture with height (%d) in excess of 'maximumImageDimensionCube' (%d)", height, properties.maximumImageDimensionCube()));
+            }
+        }
+        // no need to check 3d textures, currently B3D will throw for that
+    }
+
+    void validateAddressMode(AddressMode mode) {
+        if (!properties.knownAddressModes().contains(mode)) {
+            throw new IllegalArgumentException(String.format("Unknown by device AddressMode (%s) provided, known formats are %s", mode.name(), Arrays.toString(properties.knownAddressModes().toArray())));
+        }
+    }
+
+    public void validatePipeline(RenderPipeline pipeline) {
+        if (pipeline.getColorLogic() != LogicOp.NONE && enabledFeatures.logicOp()) {
+            throw new IllegalArgumentException(String.format("Cannot use LogicOp other than NONE without enabling 'logicOp' device feature, %s", pipeline.getLocation()));
+        }
+        if (pipeline.getVertexFormat().getElements().size() > properties.maximumVertexInputAttributes()) {
+            throw new IllegalArgumentException(String.format("Cannot use more than %d vertex input attributes, %s attempted to use %d attributes", properties.maximumVertexInputAttributes(), pipeline.getLocation(), pipeline.getVertexFormat().getElements().size()));
+        }
+        int ubos = 0;
+        int utbs = 0;
+        final int samplers = pipeline.getSamplers().size();
+        for (RenderPipeline.UniformDescription uniform : pipeline.getUniforms()) {
+            if (!properties.knownUniformTypes().contains(uniform.type())) {
+                throw new IllegalArgumentException(String.format("Unknown by device UniformType (%s) provided for uniform %s, known types are %s", uniform.type().name(), uniform.name(), Arrays.toString(properties.knownUniformTypes().toArray())));
```

Comment:
```suggestion
                throw new IllegalArgumentException(String.format("Unknown by device UniformType (%s) provided for uniform %s, known types are %s", uniform.type().name(), uniform.name(), properties.knownUniformTypes()));
```
Converting the list or set to an array and then a string provides no benefit, same applies to all other cases where this is done.

Reply:
It does here actually, this is specifically to ensure that the enum names are printed as an array, rather than the set object's hash.
the message currently prints something like `Unknown by device UniformType (SSBO) provided for uniform mySSBO, known types are [ UBO, UTB ]`

Reply:
`Set` and `EnumSet` implement `toString()` by virtue of `AbstractCollection` doing so and `Collections.UnmodifiableCollection` (which `Collections.UnmodifiableSet` extends) forwards the call to the inner set. There is no difference in output between using the set directly and converting it to an array and then a string.

Reply:
i see, don't know how i missed that before. you should have write access to this branch if you want to fix it yourself, or i can get to it tonight.